File: typescript/tests/gpu-invariants.test.ts
/**
 * GPU pricing property invariants — cost spec §10.3.
 * Mirrors python/tests/test_gpu_invariants.py.
 */

import { describe, it, expect, beforeEach } from "vitest";
import Decimal from "decimal.js";
import {
  GpuPricingEngine,
  _resetWarningStateForTests,
} from "../src/pricing/gpu-pricing.js";
import type { CloudEnv } from "../src/cloud-detect.js";

function base(billingModel: string, overrides: Record<string, any> = {}): Record<string, any> {
  return {
    billing_model: billingModel,
    gpu_vendor: "nvidia",
    gpu_sku: "h100-80gb-sxm5",
    gpu_count: 1,
    region: null,
    duration_ms: 1000,
    gpu_seconds_used: 1.0,
    instance_type: null,
    vgpu_profile: null,
    mig_profile: null,
    ...overrides,
  };
}

function cloud(
  provider: string | null,
  region: string | null,
  source: "env" | "imds" | "dmi" | "none",
  instanceType?: string | null,
): CloudEnv {
  return { provider, region, source, instanceType: instanceType ?? null };
}

const ALL_BILLING_MODELS = [
  "per_gpu_second_active",
  "per_instance_hour",
  "per_gpu_hour_reserved",
  "per_vgpu_hour",
];

function configureFor(billingModel: string): {
  details: Record<string, any>;
  env: CloudEnv;
} {
  const details = base(billingModel);
  if (billingModel === "per_instance_hour") {
    details.instance_type = "p5.48xlarge";
    details.region = "us-east-1";
    return {
      details,
      env: cloud("aws", "us-east-1", "imds", "p5.48xlarge"),
    };
  }
  if (billingModel === "per_vgpu_hour") {
    details.instance_type = "Standard_NV6ads_A10_v5";
    details.region = "eastus";
    details.gpu_sku = "a10-vgpu-1of6";
    return {
      details,
      env: cloud("azure", "eastus", "imds", "Standard_NV6ads_A10_v5"),
    };
  }
  if (billingModel === "per_gpu_hour_reserved") {
    return { details, env: cloud("lambda_labs", null, "dmi") };
  }
  return { details, env: cloud("modal", null, "env") };
}

describe("GPU pricing property invariants", () => {
  let engine: GpuPricingEngine;
  beforeEach(() => {
    _resetWarningStateForTests();
    engine = new GpuPricingEngine();
  });

  it.each(ALL_BILLING_MODELS)(
    "Invariant 1 — costUsd >= 0 always (%s)",
    (billingModel) => {
      const { details, env } = configureFor(billingModel);
      const cost = engine.resolveGpuCost(details, env, new Decimal("1"));
      expect(cost.costUsd.greaterThanOrEqualTo(0)).toBe(true);
    },
  );

  it.each([1, 2, 5, 10])(
    "Invariant 3 — linearity in gpu_seconds_used (scale %i)",
    (scale) => {
      const env = cloud("modal", null, "env");
      const base1 = base("per_gpu_second_active", {
        gpu_seconds_used: 1.0,
        duration_ms: 1000,
      });
      const baseN = base("per_gpu_second_active", {
        gpu_seconds_used: scale,
        duration_ms: scale * 1000,
      });
      const cost1 = engine.resolveGpuCost(base1, env, new Decimal(1));
      const costN = engine.resolveGpuCost(baseN, env, new Decimal(scale));
      expect(costN.costUsd.equals(cost1.costUsd.times(scale))).toBe(true);
    },
  );

  it("Invariant 4 — H100 more expensive than A100 on Modal", () => {
    const env = cloud("modal", null, "env");
    const h100 = base("per_gpu_second_active", { gpu_sku: "h100-80gb-sxm5" });
    const a100 = base("per_gpu_second_active", { gpu_sku: "a100-80gb-sxm4" });
    const h = engine.resolveGpuCost(h100, env, new Decimal(1));
    const a = engine.resolveGpuCost(a100, env, new Decimal(1));
    expect(h.costUsd.greaterThan(a.costUsd)).toBe(true);
  });

  it("Invariant 5 — Modal per-second × 3600 within 0.5×–3× Lambda per-hour", () => {
    const modal = base("per_gpu_second_active", {
      gpu_sku: "h100-80gb-sxm5",
      gpu_seconds_used: 3600.0,
      duration_ms: 3_600_000,
    });
    const lambda = base("per_gpu_hour_reserved", {
      gpu_sku: "h100-80gb-sxm5",
      gpu_count: 1,
      gpu_seconds_used: 3600.0,
      duration_ms: 3_600_000,
    });
    const modalCost = engine.resolveGpuCost(
      modal,
      cloud("modal", null, "env"),
      new Decimal(3600),
    );
    const lambdaCost = engine.resolveGpuCost(
      lambda,
      cloud("lambda_labs", null, "dmi"),
      new Decimal(3600),
    );
    expect(modalCost.costUsd.greaterThan(0)).toBe(true);
    expect(lambdaCost.costUsd.greaterThan(0)).toBe(true);
    const ratio = modalCost.costUsd.dividedBy(lambdaCost.costUsd);
    expect(ratio.greaterThan(new Decimal("0.5"))).toBe(true);
    expect(ratio.lessThan(new Decimal("3.0"))).toBe(true);
  });

  it.each(ALL_BILLING_MODELS)(
    "Invariant 6 — confidence in {computed, estimated} (%s)",
    (billingModel) => {
      const { details, env } = configureFor(billingModel);
      const cost = engine.resolveGpuCost(details, env, new Decimal("1"));
      expect(["computed", "estimated"]).toContain(cost.costConfidence);
    },
  );

  it.each(ALL_BILLING_MODELS)(
    "Invariant 7 — pricingSource starts with gpu_catalog: (%s)",
    (billingModel) => {
      const { details, env } = configureFor(billingModel);
      const cost = engine.resolveGpuCost(details, env, new Decimal("1"));
      expect(cost.pricingSource.startsWith("gpu_catalog:")).toBe(true);
    },
  );
});

File: rust/tests/gpu_invariants_test.rs
//! GPU pricing invariants — Phase 2 Task 9.
//!
//! Rust port of `python/tests/test_gpu_invariants.py` (commit d42cc81).
//! Property tests pin the spec §10.3 invariants across all four GPU
//! billing models.
//!
//! 7 invariants (one skipped — signal events are covered by the dedicated
//! observability test file):
//! 1. cost_usd >= 0 across all 4 billing models
//! 3. Linearity in gpu_seconds_used on per_gpu_second_active
//! 4. H100 > A100 rate on Modal — newer/faster GPUs cost more
//! 5. Per-GPU-second × 3600 within 0.5-3.0x of per-GPU-hour rate on the
//!    SAME canonical SKU (serverless markup is real and intentional)
//! 6. cost_confidence ∈ {computed, estimated} — never unknown on
//!    well-formed input
//! 7. pricing_source starts with "gpu_catalog:" across all 4 models

use rust_decimal::Decimal;
use rust_decimal_macros::dec;
use serde_json::json;

use dexcost::cloud_detect::CloudEnv;
use dexcost::pricing::gpu_pricing::GpuPricingEngine;

fn env(provider: Option<&str>, region: Option<&str>, itype: Option<&str>) -> CloudEnv {
    CloudEnv {
        provider: provider.map(String::from),
        region: region.map(String::from),
        source: "test",
        instance_type: itype.map(String::from),
    }
}

#[test]
fn invariant_1_cost_usd_non_negative_across_all_billing_models() {
    let e = GpuPricingEngine::new();
    let cases = vec![
        (
            json!({
                "billing_model": "per_gpu_second_active",
                "gpu_seconds_used": "10", "gpu_sku": "h100-80gb-sxm5",
                "duration_ms": 10000, "gpu_count": 1,
            }),
            env(Some("modal"), None, None),
        ),
        (
            json!({
                "billing_model": "per_instance_hour",
                "gpu_seconds_used": "30", "gpu_count": 8,
                "duration_ms": 60000, "region": "us-east-1",
                "instance_type": "p5.48xlarge",
            }),
            env(Some("aws"), Some("us-east-1"), Some("p5.48xlarge")),
        ),
        (
            json!({
                "billing_model": "per_gpu_hour_reserved",
                "gpu_seconds_used": "60", "gpu_count": 8,
                "duration_ms": 60000, "gpu_sku": "h100-80gb-sxm5",
            }),
            env(Some("lambda_labs"), None, None),
        ),
        (
            json!({
                "billing_model": "per_vgpu_hour",
                "gpu_seconds_used": "10", "gpu_count": 1,
                "duration_ms": 10000, "region": "eastus",
                "instance_type": "Standard_NV6ads_A10_v5",
            }),
            env(Some("azure"), Some("eastus"), Some("Standard_NV6ads_A10_v5")),
        ),
    ];
    for (details, ce) in cases {
        let c = e.resolve_gpu_cost(&details, &ce, None);
        assert!(
            c.cost_usd >= Decimal::ZERO,
            "billing_model={} produced negative cost: {}",
            details.get("billing_model").unwrap(),
            c.cost_usd
        );
    }
}

#[test]
fn invariant_3_linearity_in_gpu_seconds_used_modal_per_gpu_second() {
    // cost(n * gpu_seconds) == n * cost(gpu_seconds) on per_gpu_second_active.
    let e = GpuPricingEngine::new();
    let base_seconds: i64 = 1;
    let base_details = json!({
        "billing_model": "per_gpu_second_active",
        "gpu_seconds_used": base_seconds.to_string(),
        "gpu_sku": "h100-80gb-sxm5",
        "duration_ms": 1000,
        "gpu_count": 1,
    });
    let ce = env(Some("modal"), None, None);
    let c1 = e.resolve_gpu_cost(&base_details, &ce, None);

    for n in [1, 2, 5, 10] {
        let det = json!({
            "billing_model": "per_gpu_second_active",
            "gpu_seconds_used": (base_seconds * n).to_string(),
            "gpu_sku": "h100-80gb-sxm5",
            "duration_ms": (1000 * n),
            "gpu_count": 1,
        });
        let cn = e.resolve_gpu_cost(&det, &ce, None);
        let expected = c1.cost_usd * Decimal::from(n);
        let diff = (cn.cost_usd - expected).abs();
        assert!(
            diff < dec!(0.0001),
            "linearity violated at n={}: cn={} expected={} diff={}",
            n,
            cn.cost_usd,
            expected,
            diff
        );
    }
}

#[test]
fn invariant_4_h100_rate_greater_than_a100_on_modal() {
    let e = GpuPricingEngine::new();
    let ce = env(Some("modal"), None, None);
    let h100 = json!({
        "billing_model": "per_gpu_second_active",
        "gpu_seconds_used": "1", "gpu_sku": "h100-80gb-sxm5",
        "duration_ms": 1000, "gpu_count": 1,
    });
    let a100 = json!({
        "billing_model": "per_gpu_second_active",
        "gpu_seconds_used": "1", "gpu_sku": "a100-80gb",
        "duration_ms": 1000, "gpu_count": 1,
    });
    let ch = e.resolve_gpu_cost(&h100, &ce, None);
    let ca = e.resolve_gpu_cost(&a100, &ce, None);
    // Either both resolve from catalog (then h100 > a100), or both fall
    // through hardcoded (then equal). Newer/faster GPUs must NEVER cost
    // less than older/slower ones at the same provider.
    assert!(
        ch.cost_usd >= ca.cost_usd,
        "H100 rate must be >= A100 on Modal — h100={} a100={}",
        ch.cost_usd,
        ca.cost_usd
    );
}

#[test]
fn invariant_5_per_second_x_3600_close_to_per_hour_on_same_sku() {
    // Modal serverless H100 vs Lambda Labs reserved H100 — the gap IS the
    // point of the catalog. Allow a generous 0.5x to 3.0x band (serverless
    // markup is real but bounded).
    let e = GpuPricingEngine::new();
    let modal = e.resolve_gpu_cost(
        &json!({
            "billing_model": "per_gpu_second_active",
            "gpu_seconds_used": "3600",
            "gpu_sku": "h100-80gb-sxm5",
            "duration_ms": 3600000,
            "gpu_count": 1,
        }),
        &env(Some("modal"), None, None),
        None,
    );
    let lambda = e.resolve_gpu_cost(
        &json!({
            "billing_model": "per_gpu_hour_reserved",
            "gpu_seconds_used": "3600",
            "gpu_count": 1,
            "duration_ms": 3600000,
            "gpu_sku": "h100-80gb-sxm5",
        }),
        &env(Some("lambda_labs"), None, None),
        None,
    );
    if modal.cost_usd > Decimal::ZERO && lambda.cost_usd > Decimal::ZERO {
        let ratio = modal.cost_usd / lambda.cost_usd;
        assert!(
            ratio >= dec!(0.5) && ratio <= dec!(5.0),
            "Modal vs Lambda H100 ratio out of band: modal={} lambda={} ratio={}",
            modal.cost_usd,
            lambda.cost_usd,
            ratio
        );
    }
}

#[test]
fn invariant_6_cost_confidence_never_unknown_on_well_formed_input() {
    let e = GpuPricingEngine::new();
    let cases = vec![
        (
            json!({
                "billing_model": "per_gpu_second_active",
                "gpu_seconds_used": "1", "gpu_sku": "h100-80gb-sxm5",
                "duration_ms": 1000, "gpu_count": 1,
            }),
            env(Some("modal"), None, None),
        ),
        (
            json!({
                "billing_model": "per_instance_hour",
                "gpu_seconds_used": "1", "gpu_count": 1,
                "duration_ms": 1000, "region": "us-east-1",
                "instance_type": "p5.48xlarge",
            }),
            env(Some("aws"), Some("us-east-1"), Some("p5.48xlarge")),
        ),
        (
            json!({
                "billing_model": "per_gpu_hour_reserved",
                "gpu_seconds_used": "1", "gpu_count": 1,
                "duration_ms": 1000, "gpu_sku": "h100-80gb-sxm5",
            }),
            env(Some("lambda_labs"), None, None),
        ),
        (
            json!({
                "billing_model": "per_vgpu_hour",
                "gpu_seconds_used": "1", "gpu_count": 1,
                "duration_ms": 1000, "region": "eastus",
                "instance_type": "Standard_NV6ads_A10_v5",
            }),
            env(Some("azure"), Some("eastus"), Some("Standard_NV6ads_A10_v5")),
        ),
    ];
    for (det, ce) in cases {
        let c = e.resolve_gpu_cost(&det, &ce, None);
        assert!(
            c.cost_confidence == "computed" || c.cost_confidence == "estimated",
            "billing_model={} produced cost_confidence={} — must be computed|estimated",
            det.get("billing_model").unwrap(),
            c.cost_confidence
        );
    }
}

#[test]
fn invariant_7_pricing_source_starts_with_gpu_catalog() {
    // Convention §3 — every well-formed input produces a pricing_source
    // starting with "gpu_catalog:".
    let e = GpuPricingEngine::new();
    let cases = [
        (
            json!({"billing_model": "per_gpu_second_active",
                   "gpu_seconds_used": "1", "gpu_sku": "h100-80gb-sxm5",
                   "duration_ms": 1000, "gpu_count": 1}),
            env(Some("modal"), None, None),
        ),
        (
            json!({"billing_model": "per_instance_hour",
                   "gpu_seconds_used": "1", "gpu_count": 1,
                   "duration_ms": 1000, "region": "us-east-1",
                   "instance_type": "p5.48xlarge"}),
            env(Some("aws"), Some("us-east-1"), Some("p5.48xlarge")),
        ),
        (
            json!({"billing_model": "per_gpu_hour_reserved",
                   "gpu_seconds_used": "1", "gpu_count": 1,
                   "duration_ms": 1000, "gpu_sku": "h100-80gb-sxm5"}),
            env(Some("lambda_labs"), None, None),
        ),
        (
            json!({"billing_model": "per_vgpu_hour",
                   "gpu_seconds_used": "1", "gpu_count": 1,
                   "duration_ms": 1000, "region": "eastus",
                   "instance_type": "Standard_NV6ads_A10_v5"}),
            env(Some("azure"), Some("eastus"), Some("Standard_NV6ads_A10_v5")),
        ),
    ];
    for (det, ce) in cases.iter() {
        let c = e.resolve_gpu_cost(det, ce, None);
        assert!(
            c.pricing_source.starts_with("gpu_catalog:"),
            "pricing_source must start with gpu_catalog: — got {}",
            c.pricing_source
        );
    }
}
